Issue #32: added h5 loaders and their respective config files

s-sd opened the pull request:
'paired labeled', 'paired unlabeled', 'unpaired labeled' and 'unpaired unlabeled' loaders have been added

NMontanaBrown commented on deepreg/data/h5/h5_paired_labeled_loader.py:
```diff
@@ -0,0 +1,64 @@
+import os
```
Missing module docstring

NMontanaBrown commented on deepreg/data/h5/h5_paired_unlabeled_loader.py:
```diff
@@ -0,0 +1,59 @@
+import os
```
Missing module docstring

NMontanaBrown commented on deepreg/data/h5/h5_unpaired_labeled_loader.py:
```diff
@@ -0,0 +1,60 @@
+import os
```
Missing module docstrint

NMontanaBrown commented on deepreg/data/h5/h5_unpaired_unlabeled_loader.py:
```diff
@@ -0,0 +1,49 @@
+import os
```
Missing module docstring

s-sd commented:
@NMontanaBrown 
I have made some changes in the latest commit, it would be great if you could have another look

NMontanaBrown commented:
@s-sd 
Thanks for pinging!
1. Do you think you could expand on the class docstrings on functionality and inheritance from other loaders? That way we don't have to create a new ticket for improving the docs at a later point.
2. If I understood correctly, there were some tests written for this..?  If possible add the 24 images used to test locally as .h5 so we can assert functionality?

s-sd commented:
@NMontanaBrown 

I'll expand on the functionality and inheritance in the class docstrings and then make another commit.

The tests that I did, very informally tested the H5FileLoader in a script and the generators themselves were tested by running the training and prediction, in a similar way to: test/output/test_train_predict.py
I was planning on adding the new loaders to this script so they could be included in this test but the test requires data. The current nifti data will be reduced and then I will convert to h5 after which I can create a new pull request once I have done this and in the same pull request I'll add the new loaders to test_train_predict.py script and upload the tidied up tests for the H5FileLoader. Or would it be better to include some data and the tests now and then replace it with the other data later on?

s-sd commented:
@NMontanaBrown 

I've updated the class docstrings in the latest commit

NMontanaBrown commented:
@s-sd 
Could you both unit test the loaders, and test with integration as you mentioned with test/output/test_train_predict.py? It will involve creating pared down h5 from nifti but that way the whole test thread is tied together with this and we can assert functionality before usage where it might break something else.

mathpluscode commented:
The support of nifti loader for unpaired and unlabeled is not ready yet.
After this, I feel that many functions in H5 loader are duplicated, we can move some functions to be nifti/h5 independent to simplify the code.
We can do this in a separate PR ;)

s-sd commented:
@NMontanaBrown 

Added the unit test for the file loader, added the test data and added h5 loaders to test_train_predict.py in the latest commit. It would be great if you could take a look.

NMontanaBrown commented:
@s-sd the tests are not passing CI, probably due to incorrect paths to data called which allows for it to pass locally but not on the server. Can you fix this?

s-sd commented:
@NMontanaBrown 

It looks like Yunguan made a few changes which were merged before this PR. And when I merged the master with this branch, the h5 loaders did not address the changes made. It is an easy fix (just the addition of one argument to all the loaders) and I will make commits tomorrow to address these changes. 

I will also wait for Yunguan's latest PR #64 to be merged so that I can address the changes made there without needing a new PR.

s-sd commented:
@NMontanaBrown , I've added new commits which have fixed the data structure problem and added the output and unit tests. I have tried to solve most of the linting issues but there are some lines that are too long that I have left because they were either strings inside a function or they could not be cut down due to some other reason. CI has also passed for the latest commit. It would be great if you could take a look now
